pkg/storage/secret/encrypted_value_store.go

```diff
@@ -0,0 +1,132 @@
+package secret
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/grafana/grafana/pkg/infra/db"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+var (
+	ErrEncryptedValueNotFound = errors.New("encrypted value not found")
+)
+
+type EncryptedValueStorage interface {
+	Create(ctx context.Context, encryptedValue *EncryptedValue) (*EncryptedValue, error)
+	Update(ctx context.Context, newEncryptedValue *EncryptedValue) (*EncryptedValue, error)
+	Get(ctx context.Context, uid string) (*EncryptedValue, error)
+	Delete(ctx context.Context, uid string) error
+}
+
+func ProvideEncryptedValueStorage(db db.DB, cfg *setting.Cfg, features featuremgmt.FeatureToggles) (EncryptedValueStorage, error) {
+	if !features.IsEnabledGlobally(featuremgmt.FlagGrafanaAPIServerWithExperimentalAPIs) ||
+		!features.IsEnabledGlobally(featuremgmt.FlagSecretsManagementAppPlatform) {
+		return &encryptedValStorage{}, nil
+	}
+
+	return &encryptedValStorage{db: db}, nil
+}
+
+type encryptedValStorage struct {
+	db db.DB
+}
+
+func (s *encryptedValStorage) Create(ctx context.Context, encryptedValue *EncryptedValue) (*EncryptedValue, error) {
+	encryptedValue.UID = uuid.New().String()
+	encryptedValue.Created = time.Now().Unix()
+	encryptedValue.Updated = encryptedValue.Created
+
+	err := s.db.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		if _, err := sess.Insert(encryptedValue); err != nil {
+			return fmt.Errorf("insert row: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("db failure: %w", err)
+	}
+
+	return encryptedValue, nil
+}
+
+func (s *encryptedValStorage) Update(ctx context.Context, newEncryptedValue *EncryptedValue) (*EncryptedValue, error) {
+	// First find if the encrypted value exists
+	encryptedValueRow := &EncryptedValue{UID: newEncryptedValue.UID}
+	err := s.db.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		found, err := sess.Get(encryptedValueRow)
+		if err != nil {
+			return fmt.Errorf("failed to get row: %w", err)
+		}
+
+		if !found {
+			return ErrEncryptedValueNotFound
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("db failure: %w", err)
+	}
+
+	// Update the encrypted value
+	newEncryptedValue.Updated = time.Now().Unix()
+	err = s.db.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
```

Comment:
I don't know if we care but I believe it is possible for this update to succeed if the row is deleted just before the update (depends on how `sess.Update` is implemented)

Reply:
I think we can check the rowsAffected and if it is 0, we could return a `ErrEncryptedValueNotFound`.